File: question/200NumberOfIslands.cpp
//simple BFS.
class Solution {
public:
    void bfs(vector<vector<char>> &grid, int i, int j){
        grid[i][j] = '0';
        if(i>0 && grid[i-1][j]=='1'){
            bfs(grid, i-1, j);
        }
        if(i<grid.size()-1 && grid[i+1][j]=='1'){
            bfs(grid, i+1, j);
        }
        if(j>0 && grid[i][j-1]=='1'){
            bfs(grid, i, j-1);
        }
        if(j<grid[0].size()-1 && grid[i][j+1]=='1'){
            bfs(grid, i, j+1);
        }
    }
    
    int numIslands(vector<vector<char>>& grid) {
        if(grid.size()==0) return 0;
        if(grid[0].size()==0) return 0;
        
        int count = 0;
        for(int i=0;i<grid.size();i++){
            for(int j=0;j<grid[0].size();j++){
                if(grid[i][j]=='1'){
                    bfs(grid, i, j);
                    count ++;
                }
            }
        }
        return count;
    }  
};

File: Algo/Leetcode/200NumberOfIslands.py
import sys
sys.setrecursionlimit(1000000)

class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        n = len(grid)
        m = len(grid[0])

        count = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j] == "1":
                    count += 1
                    self.dfs(grid, i, j, n, m)

        return count

    def dfs(self, grid, i, j, n, m):
        if i < 0 or i >= n or j < 0 or j >= m or grid[i][j] == "0":
            return

        grid[i][j] = "0"
        self.dfs(grid, i+1, j, n, m)
        self.dfs(grid, i-1, j, n, m)
        self.dfs(grid, i, j+1, n, m)
        self.dfs(grid, i, j-1, n, m)